core_types/src/eav.rs

```diff
@@ -122,88 +176,117 @@ pub trait EntityAttributeValueStorage: objekt::Clone + Send + Sync + Debug {
         entity: Option<Entity>,
         attribute: Option<Attribute>,
         value: Option<Value>,
-    ) -> Result<HashSet<EntityAttributeValue>, HolochainError>;
+        index_query: IndexQuery,
+    ) -> Result<BTreeSet<EntityAttributeValueIndex>, HolochainError>;
 }
 
 clone_trait_object!(EntityAttributeValueStorage);
 
 #[derive(Clone, Debug)]
-pub struct ExampleEntityAttributeValueStorageNonSync {
-    storage: HashSet<EntityAttributeValue>,
+pub struct ExampleEntityAttributeValueStorage {
+    storage: Arc<RwLock<BTreeSet<EntityAttributeValueIndex>>>,
 }
-
-impl ExampleEntityAttributeValueStorageNonSync {
-    pub fn new() -> ExampleEntityAttributeValueStorageNonSync {
-        ExampleEntityAttributeValueStorageNonSync {
-            storage: HashSet::new(),
+impl ExampleEntityAttributeValueStorage {
+    pub fn new() -> ExampleEntityAttributeValueStorage {
+        ExampleEntityAttributeValueStorage {
+            storage: Arc::new(RwLock::new(BTreeSet::new())),
         }
     }
+}
+
+pub fn increment_key_till_no_collision(
+    mut eav: EntityAttributeValueIndex,
```

Comment:
why is eav mutable?

Reply:
recursive function that returns non mutable eav. It is only mutable in the scope of this function